### DISCUSSION TITLE
Problem with updating pi point. Update_mode and buffer_mode

### DISCUSSION THREAD

**Discussion by HendrikReitsma** [Hugovdberg/PIconnect#688] [Q&A] [answered]
The following code gave me a type error:

`with PI.PIServer(server='foo') as server:
    point = server.search('foo')[0]  # the tag has to be created
    point.update_value(
        1.0,
        datetime.now(),
        UpdateMode.NO_REPLACE,
        BufferMode.BUFFER_IF_POSSIBLE,
    )`

The error:

` 279 def _update_value(self, value, update_mode, buffer_mode):
--> 280     return self.pi_point.UpdateValue(value, update_mode, buffer_mode)

TypeError: No method matches given arguments for PIPoint.UpdateValue: (<class 'OSIsoft.AF.Asset.AFValue'>, <class 'int'>, <class 'int'>)`

It seems that the update_value method receives the wrong type of update_mode and buffer_mode even though I used the default values. They are specifically expected to be enums instead of integers.

To ensure the values are enums I added the following two lines to the update_value method in PIData.py:

`update_mode = AF.Data.AFUpdateOption(int(update_mode))
        buffer_mode = AF.Data.AFBufferOption(int(buffer_mode))`

Resulting in:

`
def update_value(
        self,
        value,
        time=None,
        update_mode=UpdateMode.NO_REPLACE,
        buffer_mode=BufferMode.BUFFER_IF_POSSIBLE,
    ):
        """Update value for existing PI object.

        Args:
            value: value type should be in cohesion with PI object or
                it will raise PIException: [-10702] STATE Not Found
            time (datetime, optional): it is not possible to set future value,
                it raises PIException: [-11046] Target Date in Future.

        You can combine update_mode and time to change already stored value.
        """
        if time:
            time = to_af_time(time)
        update_mode = AF.Data.AFUpdateOption(int(update_mode))
        buffer_mode = AF.Data.AFBufferOption(int(buffer_mode))
        value = AF.Asset.AFValue(value, time)
        return self._update_value(value, update_mode, buffer_mode)
`

That seems to have fixed it and I don't get any more problems.

Does anybody have any idea why this problem was occuring for me?

**Comment by Hugovdberg** [comment, marked as answer]
Are you sure you are using the latest version of `PIconnect`? This should be fixed with version 0.10, and as you mention line numbers 279/280 for `_update_value` I think you are on version 0.9.1. It was a problem with `pythonnet` version 3.0 that no longer allowed implicit conversion from int to enum. Updating `PIconnect` to 0.10.4 should be the safest bet that you don't get these errors anymore.

**Reply by HendrikReitsma** [reply]
Thanks. Updating PIconnect to 0.10.4 solved the issue.
